### PR TITLE
[usm] extract kafka protocol

### PR DESCRIPTION
**Pull request by guyarb**
### What does this PR do?

Implements USM's kafka support as a new program inside USM.

### Motivation

We're working on a larger refactor to improve readability of USM, along with allowing enabling features easily.
Now we're addressing kafka support.

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/network/protocols/kafka/protocol.go [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+
+type protocol struct {
+	cfg            *config.Config
+	telemetry      *Telemetry
+	statkeeper     *StatKeeper
+	eventsConsumer *events.Consumer
+}
+
+const (
+	eventStream                              = "kafka"
```
Comment (DanielLavie) [inline pkg/network/protocols/kafka/protocol.go:30]: nit:
```suggestion
	eventStreamName                              = "kafka"
```

File: pkg/network/protocols/kafka/protocol.go [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+	filterTailCall                           = "socket__kafka_filter"
+	dispatcherTailCAll                       = "socket__protocol_dispatcher_kafka"
```
Comment (DanielLavie) [inline pkg/network/protocols/kafka/protocol.go:32]:
```suggestion
	dispatcherTailCall                       = "socket__protocol_dispatcher_kafka"
```

File: pkg/network/protocols/kafka/protocol.go [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+func newKafkaProtocol(cfg *config.Config) (protocols.Protocol, error) {
+	if !cfg.EnableKafkaMonitoring {
+		return nil, nil
+	}
+
+	telemetry := NewTelemetry()
+
+	return &protocol{
+		cfg:       cfg,
+		telemetry: telemetry,
```
Comment (DanielLavie) [inline pkg/network/protocols/kafka/protocol.go:74]:
```suggestion
		telemetry: NewTelemetry(),
```

File: pkg/network/protocols/kafka/protocol.go [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+	}, nil
+}
+
+// ConfigureOptions add the necessary options for the http monitoring to work,
+// to be used by the manager. These are:
+// - Set the `http_in_flight` map size to the value of the `max_tracked_connection` configuration variable.
```
Comment (DanielLavie) [inline pkg/network/protocols/kafka/protocol.go:80]: Fix comment

File: pkg/network/protocols/kafka/protocol.go [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+}
+
+func (p *protocol) PreStart(mgr *manager.Manager) (err error) {
+	p.eventsConsumer, err = events.NewConsumer(
+		eventStream,
+		mgr,
+		p.processKafka,
+	)
+	if err != nil {
+		return
```
Comment (DanielLavie) [inline pkg/network/protocols/kafka/protocol.go:99]: Why not return err?

File: pkg/network/protocols/kafka/protocol.go [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+	return nil
+}
+
+func (p *protocol) Stop(_ *manager.Manager) {
+	if p.eventsConsumer != nil {
+		p.eventsConsumer.Stop()
+	}
+}
+
+func (p *protocol) DumpMaps(_ *strings.Builder, _ string, _ *ebpf.Map) {}
```
Comment (DanielLavie) [inline pkg/network/protocols/kafka/protocol.go:118]: What about the kafkaLastTCPSeqPerConnectionMap map?

File: pkg/network/protocols/kafka/protocol.go [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+
+func (p *protocol) processKafka(data []byte) {
+	tx := (*EbpfTx)(unsafe.Pointer(&data[0]))
+	p.telemetry.Count(tx)
+	p.statkeeper.Process(tx)
+}
+
+// GetStats returns a map of HTTP stats stored in the following format:
```
Comment (DanielLavie) [inline pkg/network/protocols/kafka/protocol.go:126]:
```suggestion
// GetStats returns a map of Kafka stats stored in the following format:
```

File: pkg/network/state_test.go [hunk 8]
```diff
@@ -1955,37 +1961,37 @@ func TestKafkaStatsWithMultipleClients(t *testing.T) {
...
-	delta = state.GetDelta(client1, latestEpochTime(), nil, nil, nil, nil, getStats("my-topic"))
+	delta = state.GetDelta(client1, latestEpochTime(), nil, nil, getStats("my-topic"), nil)
 	assert.Len(t, delta.Kafka, 1)
 
 	// And the second client
-	delta = state.GetDelta(client2, latestEpochTime(), nil, nil, nil, nil, getStats("my-topic2"))
+	delta = state.GetDelta(client2, latestEpochTime(), nil, nil, getStats("my-topic2"), nil)
 	assert.Len(t, delta.Kafka, 2)
 
 	// Verify that the third client also accumulated both new HTTP stats
```
Comment (DanielLavie) [inline pkg/network/state_test.go:1993]:
```suggestion
	// Verify that the third client also accumulated new Kafka stats
```

File: pkg/network/usm/kafka_monitor_test.go [hunk 9]
```diff
@@ -380,23 +409,39 @@ func (i *PrintableInt) String() string {
 	return fmt.Sprintf("%d", *i)
 }
 
+func (i *PrintableInt) Load() int {
```
Comment (DanielLavie) [inline pkg/network/usm/kafka_monitor_test.go:412]: Why we need a PrintableInt?

**Review by DanielLavie — CHANGES_REQUESTED** [review]

File: pkg/network/protocols/kafka/protocol.go (hunk 5 above)
Comment (guyarb) [reply pkg/network/protocols/kafka/protocol.go:99]: named return

File: pkg/network/usm/kafka_monitor_test.go (hunk 9 above)
Comment (guyarb) [reply pkg/network/usm/kafka_monitor_test.go:412]: To print the actual value when `require.Eventually` fails

File: pkg/network/protocols/kafka/protocol.go (hunk 1 above)
Comment (guyarb) [reply pkg/network/protocols/kafka/protocol.go:30]: done

File: pkg/network/protocols/kafka/protocol.go (hunk 2 above)
Comment (guyarb) [reply pkg/network/protocols/kafka/protocol.go:32]: done

File: pkg/network/protocols/kafka/protocol.go (hunk 3 above)
Comment (guyarb) [reply pkg/network/protocols/kafka/protocol.go:74]: done

File: pkg/network/protocols/kafka/protocol.go (hunk 4 above)
Comment (guyarb) [reply pkg/network/protocols/kafka/protocol.go:80]: done

File: pkg/network/protocols/kafka/protocol.go (hunk 6 above)
Comment (guyarb) [reply pkg/network/protocols/kafka/protocol.go:118]: what's the benefit from that map?

File: pkg/network/protocols/kafka/protocol.go (hunk 7 above)
Comment (guyarb) [reply pkg/network/protocols/kafka/protocol.go:126]: done

File: pkg/network/state_test.go (hunk 8 above)
Comment (guyarb) [reply pkg/network/state_test.go:1993]: done

File: pkg/network/usm/kafka_monitor_test.go (hunk 9 above)
Comment (DanielLavie) [reply pkg/network/usm/kafka_monitor_test.go:412]: Why not using %d?

**Review by DanielLavie — APPROVED** [review]

File: pkg/network/protocols/kafka/protocol.go [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+		cfg:       cfg,
+		telemetry: NewTelemetry(),
+	}, nil
+}
+
+// ConfigureOptions add the necessary options for the http monitoring to work,
+// to be used by the manager. These are:
+// - Set the `kafka_last_tcp_seq_per_connection` map size to the value of the `max_tracked_connection` configuration variable.
+//
+// We also configure the http event stream with the manager and its options.
```
Comment (Yumasi) [inline pkg/network/protocols/kafka/protocol.go:80]:
```suggestion
// ConfigureOptions add the necessary options for the kafka monitoring to work,
// to be used by the manager. These are:
// - Set the `kafka_last_tcp_seq_per_connection` map size to the value of the `max_tracked_connection` configuration variable.
//
// We also configure the kafka event stream with the manager and its options.
```

**Review by hmahmood — APPROVED** [review]

**Review by val06 — APPROVED** [review]
Reviewed the ebpf-platform related files

**Review by derekwbrown — APPROVED** [review]

**Pull request merged**